modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxSavepointsTransactionalCacheTest.java

```diff
@@ -0,0 +1,562 @@
+
+        return cfg;
+    }
+
+    /** Override this method to use different cache modes for tests.*/
+    protected abstract CacheMode cacheMode();
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testGet() throws Exception {
+        final IgniteCache<Integer, Integer> cache = grid(2).getOrCreateCache(getConfig());
+
+        for (TransactionConcurrency concurrency : TransactionConcurrency.values())
+            for (TransactionIsolation isolation : TransactionIsolation.values()) {
+                try (Transaction tx = grid(2).transactions().txStart(concurrency, isolation)) {
+                    info(concurrency + " " + isolation);
+
+                    tx.savepoint("sp");
+
+                    assertEquals("Broken savepoint in " + concurrency + " " + isolation +
+                        " transaction.", null, cache.get(1));
+
+                    tx.rollbackToSavepoint("sp");
+
+                    IgniteInternalFuture<Boolean> fut = GridTestUtils.runAsync(new Callable<Boolean>() {
+                        @Override public Boolean call() throws Exception {
+                            return cache.putIfAbsentAsync(1, 1).get();
+                        }
+                    },cacheMode().name()+"_get");
+
+                    assertTrue(fut.get(3_000));
+
+                    assertEquals("Broken multithreaded rollback to savepoint in " + concurrency +
+                        " " + isolation + " transaction.", (Integer) 1, cache.get(1));
+
+                    tx.commit();
+                } finally {
+                    cache.remove(1);
+                }
+            }
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testPut() throws Exception {
+        final IgniteCache<Integer, Integer> cache = grid(2).getOrCreateCache(getConfig());
+
+        GridTestSafeThreadFactory factory = new GridTestSafeThreadFactory(cacheMode().name()+"_put");
+
+        for (TransactionConcurrency concurrency : TransactionConcurrency.values())
+            for (TransactionIsolation isolation : TransactionIsolation.values()) {
+                try (Transaction tx = grid(2).transactions().txStart(concurrency, isolation)) {
+                    info(concurrency + " " + isolation);
+
+                    tx.savepoint("sp");
+
+                    assertTrue(cache.putIfAbsent(1, 0));
+
+                    Thread t = factory.newThread(new Runnable() {
+                        @Override public void run() {
+                            assertTrue(cache.putIfAbsent(1, 1));
+                        }
+                    });
+
+                    t.start();
+
+                    //We need to make sure that second thread trying to get lock.
+                    if (concurrency == TransactionConcurrency.PESSIMISTIC)
+                        while (t.getState() != Thread.State.WAITING)
+                            factory.checkError();
```

Comment:
WAT!?

Reply:
You have better to wrap assertTrue(cache.putIfAbsent(1, 1)); into GridTestUtils.runMultiThreadedAsync()
 and check that it returned future, that isDone() == false(meaning second thread is trying to get lock but it hanged)